src/main/java/com/thealgorithms/sorts/MergeSortNoExtraSpace.java

```diff
@@ -1,73 +1,83 @@
 package com.thealgorithms.sorts;
 
 import java.util.Arrays;
-import java.util.Scanner;
 
-/*This code implements the mergeSort algorithm without extra space
-For understanding about mergesort visit :https://www.geeksforgeeks.org/merge-sort/
+/**
+ * Implementation of Merge Sort without using extra space for merging.
+ * This implementation performs in-place merging to sort the array of integers.
  */
 public final class MergeSortNoExtraSpace {
     private MergeSortNoExtraSpace() {
     }
 
-    public static void callMergeSort(int[] a, int n) {
-        int maxele = Arrays.stream(a).max().getAsInt() + 1;
-        mergeSort(a, 0, n - 1, maxele);
+    /**
+     * Sorts the array using in-place merge sort algorithm.
+     *
+     * @param array the array to be sorted
+     * @return the sorted array
+     */
+    public static int[] sort(int[] array) {
+        if (array.length == 0) {
+            return array;
+        }
+        int maxElement = Arrays.stream(array).max().getAsInt() + 1;
+        mergeSort(array, 0, array.length - 1, maxElement);
+        return array;
     }
 
-    public static void mergeSort(int[] a, int start, int end, int maxele) { // this function divides the array into 2 halves
+    /**
+     * Recursively divides the array into two halves, sorts and merges them.
+     *
+     * @param array  the array to be sorted
+     * @param start  the starting index of the array
+     * @param end    the ending index of the array
+     * @param maxElement the value greater than any element in the array, used for encoding
+     */
+    public static void mergeSort(int[] array, int start, int end, int maxElement) {
         if (start < end) {
-            int mid = (start + end) / 2;
-            mergeSort(a, start, mid, maxele);
-            mergeSort(a, mid + 1, end, maxele);
-            implementMergeSort(a, start, mid, end, maxele);
+            int middle = (start + end) / 2;
```

Comment:
```suggestion
            final int middle = (start + end) >>> 1;
```
I am puzzled that spotbugs, does not trigger the `IM_AVERAGE_COMPUTATION_COULD_OVERFLOW` warning here (cf. #5131).

Reply:
Really good approach, thanks.